src/04_language.md

```diff
@@ -101,6 +101,70 @@ patterns with traits).
 
 <mark>TODO</mark>: explicit good practices in error handling.
 
+The `Result` type is the preferred way of handling functions that can fail.
+A `Result` object must be tested, and never ignored.
+
+> ### Recommendation:
+> A crate can implement its own Error type, wrapping all possible errors.
+> It must be careful to make this type exception-safe (RFC 1236), and implement
+> `Error + Send + Sync + 'static` as well as `Display`.
+
+> ### Recommendation:
+> The `?` operator should be used to improve readability of code.
+> The `try!` macro should not be used.
+
+The [error-chain](https://crates.io/crates/error-chain) and
+[failure](https://crates.io/crates/failure) crates can be used to wrap errors.
+
+### Panics
+
+Explicit error handling (`Result`) should always be preferred instead of calling
+`panic`.  The cause of the error should be available, and generic errors should
+be avoided.
+
+Crates providing libraries should never use functions or instructions that can
+fail and cause the code to panic.
+
+Common patterns that can cause panics are:
+
+ - using `unwrap` or `expect`
+ - using `assert`
+ - an unchecked access to an array
+ - integer overflow (in debug mode)
+ - division by zero
+ - large allocations
+ - string formatting using `format!`
+
+> ### Recommendation:
+> Functions or instructions that can cause the code to panic at runtime should
+> be avoided.
+
+> ### Recommendation:
+> Array indexing must be properly tested, or the `.get()` method should be
+> used to return an `Option`
+
+<mark>TODO</mark> Check if the [no_panic](https://github.com/dtolnay/no-panic)
```

Comment:
The way the [`no-panic`](https://github.com/dtolnay/no-panic) crate works, is by a clever hack:

For each marked function, a special empty struct is associated, which is created before any call to the function, and leaked after the call. Why so? Because it has a destructor (`Drop::drop`) associated, which is thus only called if the function panics (stack unwinding => dropping all currently defined values).

The trick is that the Struct's`Drop::drop()` implementation references a non-existing extern function.

When the function cannot panic, the struct is always leaked and the destructor never called. With optimizations enabled, if this can be proved, dead code elimination gets rid of the (unique) call to the unreferenced extern function and all is fine. In all the other cases, however, this call is not eliminated and thus leads to a linker error.

It looks like it can actually work surprisingly well, but there should be some `#[cfg(...)]` added to the crate for builds not using the dead-code elimination optimization. Currently those should be broken.